shared/event_reader/event_parser.h

```diff
@@ -25,77 +25,179 @@
-  
-  string locals; // Any variables that the event requires to perform correctly
-  string iterdec, iterdel, iterinit, iterrm; // Overrides iterator code
-  
-  event_info();
-  event_info(string n,int i);
+    // TODO: Delete these... use {.name = name, .spelling = spelling}
+    Argument() = default;
+    Argument(std::string name_, std::string spelling_):
+        name(std::move(name_)), spelling(std::move(spelling_)) {}
+  };
+  // Any arguments to this event, using EGM spelling.
+  boost::container::small_vector<Argument, 4> arguments;
+
+  // Construct with a base EventDescriptor.
+  explicit Event(const EventDescriptor &edesc): EventDescriptor(edesc) {}
+
+  // Represents this Event as a unique string, including any parameters.
+  std::string IdString() const;
+
+  // Formats any arguments into the human name returned by EventDescriptor.
+  std::string HumanName() const;
+
+  // Returns a function name specific to this instance of an event.
+  // Identical to BaseFunctionName for non-parameterized events.
+  std::string FunctionName() const;
+
+  std::string PrefixCode() const;
+  std::string SuffixCode() const;
+
+  // Returns whether all parameters of this event are set.
+  bool IsComplete() const;
+  // This should have been obtained from the EventDescriptor.
+  // It's not applicable to a user event.
+  bool HasInsteadCode() const = delete;
+
+  std::string SubCheckFunction() const;
+  std::string SubCheckExpression() const;
+  std::string SuperCheckFunction() const;
+  std::string SuperCheckExpression() const;
+
+  bool operator==(const Event &other) const;
+  bool operator<(const Event &other) const;
+
+ private:
+  // Replaces %1 with parameter 1, %2 with parameter 2, etc.
+  std::string ParamSubst(const std::string &str) const;
 };
 
-struct main_event_info
-{
-  string name; // The name of this event. Set by "Group:", defaults to first sub event name.
-  bool is_group; // Whether or not this event 
-  map<int,event_info*> specs;
-  typedef map<int,event_info*>::iterator iter;
-  main_event_info();
+class EventData {
+ public:
+  // Look up an event by its legacy ID pair.
+  const Event get_event(int mid, int sid) const;
+  // Look up a legacy ID pair for a non-parameterized event.
+  LegacyEventPair reverse_get_event(const EventDescriptor &) const;
+  // Look up a legacy ID pair for an event.
+  LegacyEventPair reverse_get_event(const Event &) const;
+  // Retrieve the sequence of all declared events.
+  const std::vector<EventDescriptor> &events() const { return event_wrappers_; }
+
+  // Decodes an Event ID string, such as Keyboard[Left], into an Event object.
+  Event DecodeEventString(const std::string &evstring);
+
+  EventData(buffers::config::EventFile&&);
+
+ private:
+  // Raw data read from the event configuration.
+  buffers::config::EventFile event_file_;
+  // Map of type and constant name pair (eg, {"key", "enter"}) to alias info.
+  std::map<std::pair<std::string, std::string>,
```

Comment:
again, im not biggest fan of std::pair but atleast you comment what the pairs are here